lightning/src/chain/channelmonitor.rs

```diff
@@ -1234,6 +1344,107 @@ impl<Signer: Sign> ChannelMonitor<Signer> {
 	pub fn current_best_block(&self) -> BestBlock {
 		self.inner.lock().unwrap().best_block.clone()
 	}
+
+	/// Gets the balances in this channel which are either claimable by us if we were to
+	/// force-close the channel now or which are claimable on-chain or claims which are awaiting
+	/// confirmation.
+	///
+	/// Any balances in the channel which are available on-chain (ignoring on-chain fees) are
+	/// included here until an [`Event::SpendableOutputs`] event has been generated for the
+	/// balance, or until our counterparty has claimed the balance and accrued several
+	/// confirmations on the claim transaction.
+	///
+	/// See [`ClaimableBalance`] for additional details on the types of claimable balances which
+	/// may be returned here and their meanings.
+	pub fn get_claimable_balances(&self) -> Vec<ClaimableBalance> {
+		let mut res = Vec::new();
+		let us = self.inner.lock().unwrap();
+
+		let mut confirmed_txid = us.funding_spend_confirmed;
+		if let Some((txid, conf_thresh)) = us.onchain_events_awaiting_threshold_conf.iter().find_map(|event| {
+			if let OnchainEvent::FundingSpendConfirmation { txid, .. } = event.event {
+				Some((txid, event.confirmation_threshold()))
+			} else { None }
+		}) {
+			debug_assert!(us.funding_spend_confirmed.is_none(), "We have a pending funding spend awaiting confirmation, we can't have confirmed it already!");
+			confirmed_txid = Some(txid);
+			res.push(ClaimableBalance::ClaimableAwaitingConfirmations {
+				claimable_amount_satoshis: us.current_holder_commitment_tx.to_self_value_sat,
+				confirmation_height: conf_thresh,
+			});
+		}
+
+		macro_rules! walk_htlcs {
+			($holder_commitment: expr, $htlc_iter: expr) => {
+				for htlc in $htlc_iter {
+					if let Some(htlc_input_idx) = htlc.transaction_output_index {
+						if us.htlcs_resolved_on_chain.iter().any(|v| v.input_idx == htlc_input_idx) {
+								assert!(us.funding_spend_confirmed.is_some());
+						} else if htlc.offered == $holder_commitment {
+							res.push(ClaimableBalance::MaybeClaimableHTLCAwaitingTimeout {
+								claimable_amount_satoshis: htlc.amount_msat / 1000,
+								claimable_height: htlc.cltv_expiry,
+							});
+						} else {
+							if us.payment_preimages.get(&htlc.payment_hash).is_some() {
+								if let Some(conf_thresh) = us.onchain_events_awaiting_threshold_conf.iter().find_map(|event| {
+									if let OnchainEvent::HTLCSpendConfirmation { input_idx, .. } = event.event {
+										if input_idx == htlc_input_idx { Some(event.confirmation_threshold()) } else { None }
+									} else { None }
+								}) {
+									res.push(ClaimableBalance::ClaimableAwaitingConfirmations {
+										claimable_amount_satoshis: htlc.amount_msat / 1000,
+										confirmation_height: conf_thresh,
+									});
+								} else {
+									res.push(ClaimableBalance::ContentiousClaimable {
```

Comment:
Can this happen for both HTLC success and HTLC timeout, just before they appear in a block?